### ISSUE TITLE
easy personal test environments

### ISSUE THREAD

**Issue by 7yl4r** [USF-IMARS/imars_dags#85]
Testing things in airflow on the production or test server is too big of a hurdle right now. Isolated personal development environments should help with this.

I think I can use [this airflow dockerfile](https://github.com/puckel/docker-airflow) to spin up personal environments easily. However, I need to play with this & iron out some issues I foresee with the "imars_product_db" connection.

**Comment by 7yl4r** [comment]
hurdles to using this docker image:

1. need to mount NFS shares `/srv/imars-objects/`
2. also needs isolated (and pre-populated?) `imars_product_db`
3. needs custom software installs
    1. ocssw/snap
    2. imars-etl
    3. matlab

These are not at all trivial... Is it going to be easier to simplify VM cloning or port these things into docker? :thinking:

**Comment by 7yl4r** [comment]
Cloning the test VM & modifying ([w/ `virt-sysprep`](https://www.cyberciti.biz/faq/reset-a-kvm-clone-virtual-machines-with-virt-sysprep-on-linux/) or manually) is going to be way easier. Not nearly as glamorous, but way more practical.

**Comment by 7yl4r** [comment]
I think the ideal behavior we want is

* everything looks like production, but is read-only
* but attempted writes are remembered locally

This is local-i/o fallback to prod-i/o... I don't know an easy way to achieve that. :thinking: 

Parts of the airflow-test VM that are connected to (:link:) or disconnected from (:heavy_check_mark:) 
 production:

1. :link:  login user/pw
2. :heavy_check_mark: executor db (test uses localExec configured by puppet)
3. :link: imars_product_db
4. :link: imars-etl data load/extract

(1) is not a big deal and may be easily fixed w/ puppet. 

(3) is the big challenge. I think the fix is:

```
a) set up local db using schema from puppet (or maybe even backups of production?)
b) modify airflow named connection puppet
```

(4) is probably best fixed by failing all loads somehow but extracts need to stay on? Maybe (4) is the big challenge? :thinking:

**Comment by 7yl4r** [comment]
I can now see how everything can fit together...except for the networking. There are going to be issues:

* setting up static IPs for each node (maybe https://github.com/USF-IMARS/issues/issues/56 is a prereq?)
* connecting to /srv/imars-objects without IMaRS LAN

**Comment by 7yl4r** [comment]
I think I lost sight of the bigger picture and need to back up for a minute.

Puppet can set up everything within a machine. What we want here is not a process for cloning af-test, instead:

* basic outline of `data/node` data for a personal airflow test machine
* puppet `role::airflow_test_machine`? ~~or `profile::airflow_test_machine` - what roles would this go under, `linux_desktop`?~~ (edit: yeah nah role is cleaner)
* set of documents for users on how to either a) set up linux & puppet or b) use & modify a provided VM

**Comment by 7yl4r** [comment]
maybe [virt-builder](http://libguestfs.org/virt-builder.1.html) combined with a [puppet client setup script](https://github.com/USF-IMARS/IMaRS-docs/blob/master/docs/puppet/client-setup.md) is the easiest way to get this done...

```
tylar@tylardesk:~/virtbuild$ chmod a+r /boot/vmlinuz-*
tylar@tylardesk:~/virtbuild$ virt-builder -v centos-7.5 \
--run ./centos7setup.sh \
--hostname imars-airflow-tylar.marine.usf.edu \
-o airflow-tylar.img \
--size 75G
```

bah... [problems](https://askubuntu.com/questions/858966/has-anyone-successfully-run-virt-builder-on-16-04).

**Comment by 7yl4r** [comment]
more problems...

```
[root@manglillo virtbuild]# virt-builder -v centos-7.5 --run ./centos7setup.sh --hostname imars-airflow-tylar.marine.usf.edu -o ./airflow-tylar.img --size 75G
virt-resize: error: unknown/unavailable method for expanding the xfs 
filesystem on /dev/sda4

If reporting bugs, run virt-resize with debugging enabled and include the 
complete output:

  virt-resize -v -x [...]
```

**Comment by 7yl4r** [comment]
Above issue on manglillo is a [known issue with easy fix](http://libguestfs.org/virt-resize.1.html#unknown-unavailable-method-for-expanding-the-type-filesystem-on-device-lv) (`yum install -y libguestfs-xfs`).

Works now:

```
[root@manglillo virtbuild]# virt-builder -v centos-7.5 --run ./centos7setup.sh --hostname imars-airflow-tylar.marine.usf.edu -o ./airflow-tylar.img --size 75G
[...]
                   Output file: ./airflow-tylar.img
                   Output size: 75.0G
                 Output format: raw
            Total usable space: 74.4G
                    Free space: 73.1G (98%)
[root@manglillo virtbuild]# ls -lh
total 1.3G
-rw-r--r--. 1 root root 75G Oct 22 19:59 airflow-tylar.img
-rw-r--r--. 1 root root 631 Oct 17 19:04 centos7setup.sh
```

**Comment by 7yl4r** [comment]
Post-provisioning steps:

* login as root using pw generated during build
* use `passwd` to change your pw
* sign the cert on puppet master
* wait for puppet or run now w/ `puppet agent -t`
    - this will take a while (up to 1hr)
* (optional) `sytemctl enable airflow-scheduler && systemctl enable airflow-webserver && systemctl enable airflow-worker` 
* (?) `airflow initdb`
* install imars_dags dependencies `pip3 install git+https://git}@github.com/7yl4r/pycmr.git/@prod`
* switch to airflow user using `su - airflow` (you can now log in as this user too)

Workflow options:

* edit dags in `/home/airflow/dags/`
* use airflow CLI
* use airflow GUI by going to `192.168.122.*` address found in `ip a` output

**Comment by 7yl4r** [comment]
All the pieces of the puzzle are in place. Let's try setting up a env for sebastian on hydra:

```
[root@hydra ~]# yum install -y libguestfs-xfs
[...]
Installed:
  libguestfs-xfs.x86_64 1:1.36.10-6.el7_5.2                                                                                                                                                  
[...]
[root@hydra ~]# virt-builder -v centos-7.5 \
    --run /root/virt-builder-scripts/centos7setup.sh \
    --hostname imars-airflow-sebastian.marine.usf.edu \
    -o /home/images/imars-airflow-sebastian.qcow2 \
    --size 75G \
    --format qcow2 \
    --root-password password:'YadaYada'
[...]
                   Output file: /home/images/imars-airflow-sebastian.qcow2
                   Output size: 75.0G
                 Output format: qcow2
            Total usable space: 74.4G
                    Free space: 73.0G (98%)

[root@hydra ~]# virt-install \
   -n imars-airflow-sebastian \
   --os-type=Linux \
   --ram=4096 \
   --vcpus=2 \
   --disk path=/home/images/imars-airflow-sebastian.qcow2,bus=virtio,size=75 \
   --boot hd \
   --import \
   --network network=external-nat,model=virtio
WARNING  Graphics requested but DISPLAY is not set. Not running virt-viewer.
WARNING  No console to launch for the guest, defaulting to --wait -1

Starting install...
Creating domain...                                                                                                                                                                                                   |    0 B  00:00:00     
Domain installation still in progress. Waiting for installation to complete.
[...LONG WAIT...]
^C
Domain install interrupted.
Installation aborted at user request
You have new mail in /var/spool/mail/root
```

install seems to get stuck there, but it looks like it installed ok.

...although the password isn't working so I can't get in...

**Comment by 7yl4r** [comment]
tried using `--format qcow2` + `virt manager` & got "no boot device". :hankey:

**Comment by 7yl4r** [comment]
Ohh... just noticing something interesting. It is not that the password isn't working. The login session starts and then immediately terminates.

**Comment by 7yl4r** [comment]
I'm revisiting this with a fresh perspective and thinking: maybe users don't need *more control* they need a *shorter workflow loop*.

Idea: Let's re-think how users use the airflow-test server. Users are familiar with logging into a server and running commands. I could (1) allow everyone ssh access to the `airflow` account or (2) [configure sudo](https://serverfault.com/a/17818/392592) so user accounts can run things like `sudo -u airflow airflow list_dags`. (2) is cleaner, but (1) is quick and easy.

The remaining issue is: what about editing `dags` directly on the server? No one likes using a terminal editor, so what if we mount this as an NFS share and allow editing in the /homes/ dir? I am afraid this will lead to people not using git, but un-versioned use might be better than no use I guess. This also breaks the "automatic update from github" concept because `vcsrepo` squashes. we need to choose between (a) the github workflow and (b) the NFS workflow or (c) maybe I can get `vcsrepo` to not squash?

**Comment by 7yl4r** [comment]
Interesting read [here](https://medium.com/bluecore-engineering/were-all-using-airflow-wrong-and-how-to-fix-it-a56f14cb0753) that has made me rethink my approach.

Users don't need to test in airflow, airflow DAGs need to be simple enough so they require minimal testing. We aren't ready for `KubernetesOperators` but if we stick to basic `BashOperators` using the `imars-etl` CLI users can test the bash scripts outside of airflow.
